### DISCUSSION TITLE
Tumour file, case/control and error in sizeAnnotation.R step

### DISCUSSION THREAD

**Discussion by MaceJulien** [nrlab-CRUK/INVAR2#7] [Q&A]
Hello there,
I'm trying to use INVAR to detect if some patients still have ctDNA after treatment, with the potential distant goal to decide if cancer was succesfully cured or not if we can tune precisely enough parameters.
Right now, I've got an error stopping pipeline execution at what devs call Step2, in the sizeAnnotation.R script:

<details><summary>Error logs</summary>
<p>

 ! Can't subset columns that don't exist.
  ✖ Column `UNIQUE_POS` doesn't exist.
  Backtrace:
       ▆
    1. ├─global main(parseOptions())
    2. │ └─global equaliseSizeCounts(mutationsTable, fragmentSizesTable, threads = scriptArgs$THREADS)
    3. │   └─... %>% select(UNIQUE_POS, SIZE, MUTANT)
    4. ├─dplyr::select(., UNIQUE_POS, SIZE, MUTANT)
    5. ├─dplyr:::select.data.frame(., UNIQUE_POS, SIZE, MUTANT)
    6. │ └─tidyselect::eval_select(expr(c(...)), data = .data, error_call = error_call)
    7. │   └─tidyselect:::eval_select_impl(...)
    8. │     ├─tidyselect:::with_subscript_errors(...)
    9. │     │ └─rlang::try_fetch(...)
   10. │     │   └─base::withCallingHandlers(...)
   11. │     └─tidyselect:::vars_select_eval(...)
   12. │       └─tidyselect:::walk_data_tree(expr, data_mask, context_mask)
   13. │         └─tidyselect:::eval_c(expr, data_mask, context_mask)
   14. │           └─tidyselect:::reduce_sels(node, data_mask, context_mask, init = init)
   15. │             └─tidyselect:::walk_data_tree(new, data_mask, context_mask)
   16. │               └─tidyselect:::as_indices_sel_impl(...)
   17. │                 └─tidyselect:::as_indices_impl(...)
   18. │                   └─tidyselect:::chr_as_locations(x, vars, call = call, arg = arg) 
   19. │                     └─vctrs::vec_as_location(...)
   20. └─vctrs (local) `<fn>`()
   21.   └─vctrs:::stop_subscript_oob(...)
   22.     └─vctrs:::stop_subscript(...)
   23.       └─rlang::abort(...)
  Message d'avis :
  Dans main(parseOptions()) :
    There are no rows for sample 4 in mutation_table.on_target.rds
  Exécution arrêtée

</p>
</details> 

A bit more details about data that were used:

- 24 control samples, flagged as "control_negative" in the corresponding column in layout file, from 24 different individuals, without any illness nor any link with my "case" samples
- 2 patients samples, Patient1_baseline and Patient2_baseline, flagged as "case", that are ctDNA from patients with tumour (maybe I should flag them as "control_positive" instead)
- 2 patient samples, Patient1_Day30 and Patient2_Day30, flagged as "case", as you could guess 30 days after treatment, where the status is unknown and needs to be identified (Biologists already know the status by some other analysis)

What's inside the tumour file: validated SNVs from the 2 baseline samples Patient1 and Patient2. Not a single line from control samples nor from the Day30 samples. SNVs are detected on a capture panel sequencing (no WES nor WGS), same panel used for all samples.

**Something to precise: sample N°4 that have no line according to the error is a control one.**

Following this error and in order to solve it, I've got several questions about the detailed input required and how INVAR works:

1. Could the error come from the absence of SNVs related to sample N°4 in the tumour file? (that's my main hypothesis right now, but I hope I'm wrong as I do not have any SNPs list for control samples and i don't understand what would be required for controls)
2. Could it come from background filtering that is too stringent? (I don't think so as I'm using default parameters for this execution)
3. Could it come from the filtering step that is comparing 1 sample with all the others? (I don't think so as it's supposed to happen later, but I could not understand precisely the background generation so ...) => Should I just change the flag from the baseline samples / could it be the "control_negative" flag that is not well understood by the programs? (And briefly how does this step work exactly, as I could not find out?)

Thanks a lot for taking the time to read this and the time you may spend to help me!
Have a nice day
Julien

PS: I'm sorry for some french words that are in the error logs, they should not interfere with interpretation

Update 1, date July 16th 2024: I tried a rerun without the Day30 samples, I've got the same error even on the baseline samples files. It seems that as supposed, the Day30 samples do not have anything to do with this error

**Comment by ed5152** [comment]
Hi Julian, 

The lack of UNIQ_POS column is a common error that comes from having different patient names (or unique identifiers) in the layout and the mutation file. Check the debug_commonErrors doc. 
Also make sure your default parameters are appropriate for your data (ie max_depth per loci etc...) so that you don't end up filtering everything out. If you only have two patients but multiple time points for each, make sure your PROPORTION_OF_CONTROLS filter is set very high, because you expect to see signal in may other "control" samples if they all come from the same patient. 

Re your questions:
1. No SNV lists are required for control samples. The input mutation lists are called on control samples to check for noise at those loci.
2. I would double check your input parameters. The default parameters are definitely not correct for a cohort of only 2 patients.
3. Your error is coming from an issue with the inputs / enacting of the parameters. Play around with the defaults/ set everything to 0/1 as applicable to get a successful first run, then fine tune.

Good luck and let us know how it goes :)
Emma

**Reply by MaceJulien** [reply]
Hello Emma,
Thanks for your answer, you're confirming what I've suspecting after checking code for 2 days: ProportionsOfControls and MaxBackgroundAF parameters are definitely too low by default considering how it is applied and how many cases I have, as they will make the double filter status "locus_noise.both_strands" too hard to reach. The error about different IDs is the first thing I checked, as it happened to me several times already in diverse situations, but no error this time ;)


While I was reading the code, I found something that was unexpected to me in the OffTargetErrorRates.R script, line 87:
```
    layoutTable.cases <- layoutTable %>%
        filter(CASE_OR_CONTROL == 'case')
```
Why considering "case" samples and not the "control" samples when creating the lociErrorRateTable (even more, the "control_neg" samples, if someone decides to put some "control_pos" inside)? If I understood well the script goal, I'd expect it to report how much the mutation is detected in my control_neg, or at least in all considered samples, not only in "cases"

I'd like to add (I updated the original post after this message) that the 24 control samples come from 24 different individuals, without any illness nor any link with my "case" samples, in case it was not clear. But until now, I could not understand how these "controls" are used (as I was looking for the error source, I didn't go further than CreateOnTargetMutationsTable.R for now, I bet their use is later according to the docs).

**Comment by MaceJulien** [comment]
Hello,
After several weeks of trials, INVAR finally completed a successful run. Now, time to benchmark to make it well suited for experimental design described above.
For those interested, I faced several issues, including the one in the original message (corrected by Emma's suggestion just above), "infinite weights" for some samples when calculating step 4 GeneralisedLikehoodRatioTest, broken symlinks by Nextflow when trying to execute step 3 markOutliers, another one in step 5 Analysis with the splice function (described in Issues with the precise location and solution), and maybe a few others that I could not precisely locate or describe. I did not find issues causes, but I'll keep investigating when I'll have time. At least I could explain you what I did that seemed to avoid the issues if you face them.

Once I'll have satisfying parameters, I'll make a complete summary that will recap experimental design and chosen parameters.
Have a nice day
Julian
